src/wallet/wallet.cpp

```diff
@@ -1650,33 +1708,50 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
 
     {
         LOCK2(cs_main, cs_wallet);
-        for (map<uint256, CWalletTx>::const_iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
-        {
-            const uint256& wtxid = it->first;
-            const CWalletTx* pcoin = &(*it).second;
 
-            if (!CheckFinalTx(*pcoin))
+        std::set<const CWalletTx*> rejectedCoins;
+        for (TxUnspents::const_iterator uit = mapTxUnspents.begin(); uit != mapTxUnspents.end(); ++uit) {
+            const COutPoint &outpoint = uit->first;
+            const uint256& wtxid = outpoint.hash;
+            const CWalletTx* pcoin = (uit->second).first;
+
+            if(rejectedCoins.find(pcoin) != rejectedCoins.end())
                 continue;
 
-            if (fOnlyConfirmed && !pcoin->IsTrusted())
+            if (!CheckFinalTx(*pcoin)) {
+                rejectedCoins.insert(pcoin);
                 continue;
+            }
 
-            if (pcoin->IsCoinBase() && pcoin->GetBlocksToMaturity() > 0)
+            if (fOnlyConfirmed && !pcoin->IsTrusted()) {
+                rejectedCoins.insert(pcoin);
                 continue;
+            }
 
-            int nDepth = pcoin->GetDepthInMainChain();
-            if (nDepth < 0)
+            if (pcoin->IsCoinBase() && pcoin->GetBlocksToMaturity() > 0) {
+                rejectedCoins.insert(pcoin);
                 continue;
+            }
 
-            for (unsigned int i = 0; i < pcoin->vout.size(); i++) {
-                isminetype mine = IsMine(pcoin->vout[i]);
-                if (!(IsSpent(wtxid, i)) && mine != ISMINE_NO &&
-                    !IsLockedCoin((*it).first, i) && (pcoin->vout[i].nValue > 0 || fIncludeZeroValue) &&
-                    (!coinControl || !coinControl->HasSelected() || coinControl->fAllowOtherInputs || coinControl->IsSelected((*it).first, i)))
-                        vCoins.push_back(COutput(pcoin, i, nDepth,
-                                                 ((mine & ISMINE_SPENDABLE) != ISMINE_NO) ||
-                                                  (coinControl && coinControl->fAllowWatchOnly && (mine & ISMINE_WATCH_SOLVABLE) != ISMINE_NO)));
+            int nDepth = pcoin->GetDepthInMainChain();
+            if (nDepth < 0) {
+                rejectedCoins.insert(pcoin);
+                continue;
             }
+
+            if(!(pcoin->vout[outpoint.n].nValue > 0 || fIncludeZeroValue))
+                continue;
+
+            if(IsLockedCoin(wtxid, outpoint.n))
```

Comment:
This test is cheaper than `int nDepth = pcoin->GetDepthInMainChain(); if (nDepth < 0)...`